### PR TITLE
JENKINS-46253 Including support for subPath in configMaps Volumes

### PR DESCRIPTION
**Pull request by mcamposv**
- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
        JENKINS-46253 Including support for subPath in configMaps Volumes 
- [X] Link to relevant issues in GitHub or Jira
      https://issues.jenkins.io/browse/JENKINS-46253?jql=text%20~%20subPath
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by hameno** [comment]
Can this be merged soon? We also need this functionality.

**Comment by hameno** [comment]
I think this should also be expanded to the other volume types.

File: pom.xml [hunk 1]
```diff
@@ -93,6 +93,7 @@
     <dependency>
       <groupId>org.jenkins-ci.plugins</groupId>
       <artifactId>durable-task</artifactId>
+      <version>1.39</version>
```
Comment (Vlatombe) [inline pom.xml:96]: Why do you need to bump this? It looks unrelated to the change.

File: pom.xml [hunk 2]
```diff
@@ -47,7 +47,7 @@
     <jenkins.host.address />
     <slaveAgentPort />
     <java.level>8</java.level>
-    <jenkins.version>2.263.1</jenkins.version>
+    <jenkins.version>2.277.1</jenkins.version>
```
Comment (Vlatombe) [inline pom.xml:50]: I believe this is motivated by the `durable-task` bump below, but I don't think it is necessary.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java [hunk 3]
```diff
@@ -81,6 +81,7 @@
 import static org.csanchez.jenkins.plugins.kubernetes.PodTemplateUtils.combine;
 import static org.csanchez.jenkins.plugins.kubernetes.PodTemplateUtils.isNullOrEmpty;
 import static org.csanchez.jenkins.plugins.kubernetes.PodTemplateUtils.substituteEnv;
+import org.csanchez.jenkins.plugins.kubernetes.volumes.ConfigMapVolume;
```
Comment (Vlatombe) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java:84]: Should be line 50.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java [hunk 4]
```diff
@@ -173,8 +174,16 @@ public Pod build() {
...
-                        .withMountPath(mountPath).withName(volumeName).withReadOnly(false).build());
+                VolumeMountBuilder volumeMountBuilder = new VolumeMountBuilder() //
+                        .withMountPath(mountPath).withName(volumeName).withReadOnly(false);
+
+                if (volume instanceof ConfigMapVolume) {
+                    final ConfigMapVolume configmapVolume = (ConfigMapVolume) volume;
+                        //We need to normalize the subPath or we can end up in really hard to debug issues Just in case.
+                        final String subPath = substituteEnv(Paths.get(configmapVolume.getSubPath()).normalize().toString().replace("\\", "/"));
+                        volumeMountBuilder = volumeMountBuilder.withSubPath(subPath);
+                            }
```
Comment (Vlatombe) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java:185]: Please fix indentation.

File: .gitignore [hunk 5]
```diff
@@ -8,6 +8,7 @@
 /.idea/
 *.iml
 nbactions.xml
+nb-configuration.xml
```
Comment (Vlatombe) [inline .gitignore:11]: Unrelated, but ok.

File: pom.xml (hunk 2 above)
Comment (mcamposv) [reply pom.xml:50]: yes this is motivated by de durable task if we can "fix that" I have no need to bump up the Jenkins version. 
actually, I prefer to be compatible with the lowest version possible.

File: .gitignore (hunk 5 above)
Comment (mcamposv) [reply .gitignore:11]: oups my bad,  I'll fix my git ignore.

File: pom.xml (hunk 1 above)
Comment (mcamposv) [reply pom.xml:96]: maven fail to build the hpi due to dependencies. if I do not bump it.
I can try to check which one is the one that allows hpi to be build.
I'll re-check with what is on master right now. and see if it build but I think I did a pull and that is where started to fail.

Comment (mcamposv) [reply pom.xml:96]: I think that this is required because workflow-durable-task-step released a new version 2.40 that requires durable-task >1.36
so possible solutions are.
bump up durable tasks to 1.36 and its dependencies.
or limit  workflow-durable-task-step to 2.39
Both solutions are ok.
Which one do you prefer me to do?

Comment (Vlatombe) [reply pom.xml:96]: You shouldn't have to do this. Please revert the jenkins core update and this line.

Most plugin versions are managed by the jenkins bom. In general it shouldn't be changed unless there is a need for a new api or a bugfix that is affecting the plugin.

File: .gitignore (hunk 5 above)
Comment (Vlatombe) [reply .gitignore:11]: I wrote "unrelated" because it's unrelated to the change you're making in this PR. Usually it is good practice to split this kind of changes into different PRs as it makes them easier to review and to audit later in case of regressions.

Tip: It's possible to configure a global `.gitignore` for your user https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files#configuring-ignored-files-for-all-repositories-on-your-computer

**Review by Vlatombe — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 1 above)
Comment (mcamposv) [reply pom.xml:96]: I'm missing something. if I leave pom.xml as it is I've got this error.
`Failed while enforcing RequireUpperBoundDeps. The error(s) are [
Require upper bound dependencies error for org.jenkins-ci.plugins:durable-task:1.35 paths to dependency are:
+-org.csanchez.jenkins.plugins:kubernetes:1.30.3-SNAPSHOT
  +-org.jenkins-ci.plugins:durable-task:1.35
and
+-org.csanchez.jenkins.plugins:kubernetes:1.30.3-SNAPSHOT
  +-org.jenkins-ci.plugins.workflow:workflow-durable-task-step:2.40
    +-org.jenkins-ci.plugins:durable-task:1.37 (managed) <-- org.jenkins-ci.plugins:durable-task:1.36
]`

Comment (Vlatombe) [reply pom.xml:96]: That is weird, I don't have the same problem.
> +-org.csanchez.jenkins.plugins:kubernetes:1.30.3-SNAPSHOT
  +-org.jenkins-ci.plugins:durable-task:1.35

indicates you have a direct declaration of durable-task:1.35 in the pom.xml. Are you sure you don't have anything else staged?

Just in case, make sure you're running maven latest version, could be a bug with an ancient maven release.

Comment (mcamposv) [reply pom.xml]: I downloaded the last pom.xml from master branch and that works.

File: .gitignore (hunk 5 above)
Comment (mcamposv) [reply .gitignore:10]: I think I managed to resolve all the issues. can we leave the .gitignore to be included in this pr?
I really think having the NetBeans config in the git ignore is harmless and might help others.
having them in the global config file will make my life more complicated.

**Review by Vlatombe — APPROVED** [review]

**Comment by mcamposv** [comment]
Thanks
